components/salsa-2022/src/input_field.rs

```diff
@@ -0,0 +1,67 @@
+use std::hash::Hash;
+use rustc_hash::FxHashMap;
+use crate::{Durability, IngredientIndex, Revision, Runtime};
+use crate::cycle::CycleRecoveryStrategy;
+use crate::ingredient::Ingredient;
+use crate::key::DependencyIndex;
+use crate::runtime::local_state::QueryInputs;
+use crate::runtime::StampedValue;
+
+pub struct InputFieldIngredient<K, F> {
+    index: IngredientIndex,
+    map: FxHashMap<K, StampedValue<F>>
+}
+
+impl<K, F> InputFieldIngredient<K, F>
+where K: Eq + Hash
+{
+    pub fn new(index: IngredientIndex) -> Self {
+        Self {
+            index,
+            map: Default::default(),
+        }
+    }
+
+    pub fn store(
+        &mut self,
+        runtime: &mut Runtime,
+        key: K,
+        value: F,
+        durability: Durability,
+    ) -> Option<F> {
+        let revision = runtime.current_revision();
+        let stamped_value = StampedValue {
+            value,
+            durability,
+            changed_at: revision
+        };
+
+        if let Some(old_value) = self.map.insert(key, stamped_value) {
+            Some(old_value.value)
+        } else {
+            None
+        }
+    }
+
+    pub fn fetch(
+        &self,
+        key: K,
+    ) -> &F {
+        &self.map.get(&key).unwrap().value
```

Comment:
You also need to record the read here!

You need a call like this one:

https://github.com/salsa-rs/salsa/blob/8b7324dca886001df7e7bf7c45567520954fec6b/components/salsa-2022/src/function/fetch.rs#L26-L30

This will tell the active query (if any) that the function read this input.

The `database_key_index` function will also have to be added (it will read the `index` field), and should look something like this:

https://github.com/salsa-rs/salsa/blob/8b7324dca886001df7e7bf7c45567520954fec6b/components/salsa-2022/src/function.rs#L145-L150

Reply:
https://github.com/salsa-rs/salsa/pull/342/commits/9024a7bb880e1b469e07f39bfc06574ee3f99cd6 should fix this. I only passed the runtime instead of the db, that shouldn't be a problem right?